### PR TITLE
Clean up our Rubocop todos

### PR DESCRIPTION
**Pull request by jfly**
All the exceptions in our `.rubocop_todo.yml` were really bothering me, so I decided to go through and clean up as many of them as possible. This took a **ton** of time to do.

I made a bunch of executive decisions about which cops to keep or not. Comments are appreciated (I tried to keep each cop I dealt with to a separate commit), but please make your comments quickly. This branch is *begging* for conflicts. I'm going to merge this up on Monday to give people the weekend to object.

### CODE REVIEW

File: WcaOnRails/app/controllers/admin/avatars_controller.rb [hunk 1]
```diff
@@ -20,7 +20,7 @@ def update_all
           when "defer"
             # do nothing!
           else
-            raise "Unrecognized avatar action '#{args[:action]}'"
+            fail "Unrecognized avatar action '#{args[:action]}'"
```
Comment (timhabermaas) [inline WcaOnRails/app/controllers/admin/avatars_controller.rb]: Rubocop suggests different semantics for these methods: http://www.rubydoc.info/github/bbatsov/rubocop/Rubocop/Cop/Style/SignalException

This make some sense, but if everyone thinks `raise == fail` [1] then it's useless and we should just stick with `raise` everywhere.

[1] To be fair: I once knew about this Rubocop rule, but had to look it up again...

Comment (jfly) [inline WcaOnRails/app/controllers/admin/avatars_controller.rb]: Ok, good to know. Without a professional ruby background, it's hard for me distinguish between convention and "weird stuff". I often miss python with it's "one right way to do things" philosophy...

I'll change rubocop to require that we use `raise` everywhere.

File: WcaOnRails/app/controllers/competitions_controller.rb [hunk 2]
```diff
@@ -181,7 +181,7 @@ def post_results
         return
       end
 
-      top333 = comp.results.where(eventId: '333', roundId: ['f', 'c']).order(:pos).limit(3)
+      top333 = comp.results.where(eventId: '333', roundId: %w(f c)).order(:pos).limit(3)
```
Comment (jfly) [inline WcaOnRails/app/controllers/competitions_controller.rb]: Agreed, I'll remove the rule.

File: WcaOnRails/app/controllers/competitions_controller.rb [hunk 3]
```diff
@@ -238,10 +238,10 @@ def post_results
                 end
                 record_strs
               end.flatten
-              "#{uniqueName} #{record_strs.join(", ")}"
+              "#{uniqueName} #{record_strs.join(', ')}"
```
Comment (jfly) [inline WcaOnRails/app/controllers/competitions_controller.rb]: I suppose the point of this rule was to deal with poor syntax highlighters? Either way, I agree, I'll remove this.

**Comment by jfly** [comment]
> I think it's important for a project to have some kind of coding standard that's owned by the team rather than a tool, and I hope others chime in so we can get something like that. It will be different from my personal preference, and that's as it should be.

I totally agree. Part of the reason why I waited so long to do this was because I was hoping that our style would naturally evolve as people accidentally wrote code that violated rubocop's defaul rules. The rubocop authors are very clear about the fact that they intentionally turn on every cop, even if it's controversial. They encourage you to turn off or configure cops you don't like. The logic is that if the cops are off by default, you won't ever know about them.

File: WcaOnRails/app/controllers/polls_controller.rb [hunk 4]
```diff
@@ -42,11 +42,11 @@ def edit
...
-      if params[:commit] == "Confirm"
-        flash[:success] = "Poll confirmed and open to voting"
-      else
-        flash[:success] = "Updated poll"
-      end
+      flash[:success] = if params[:commit] == "Confirm"
+                          "Poll confirmed and open to voting"
+                        else
+                          "Updated poll"
+                        end
```
Comment (jfly) [inline WcaOnRails/app/controllers/polls_controller.rb]: I personally don't like the additional indentation this causes. How do you feel about turning off the cop and letting people follow their hearts with these?

File: WcaOnRails/app/controllers/polls_controller.rb [hunk 5]
```diff
@@ -70,6 +70,6 @@ def poll_params
     if params[:commit] == "Confirm" && current_user.can_create_poll?
       poll_params[:confirmed_at] = Time.now
     end
-    return poll_params
+    poll_params
```
Comment (jfly) [inline WcaOnRails/app/controllers/polls_controller.rb]: I have little opinion on this. I know there are dragons with procs and returns, and I'm okay with lowering the probability I discover those dragons.

File: WcaOnRails/app/helpers/competitions_helper.rb [hunk 6]
```diff
@@ -16,12 +16,12 @@ def competition_message_for_user(competition, user)
   end
 
   def announced_class(days_announced)
-    level = [Competition::ANNOUNCED_DAYS_WARNING, Competition::ANNOUNCED_DAYS_DANGER].select {|d| days_announced > d}.count
+    level = [Competition::ANNOUNCED_DAYS_WARNING, Competition::ANNOUNCED_DAYS_DANGER].select { |d| days_announced > d }.count
```
Comment (jfly) [inline WcaOnRails/app/helpers/competitions_helper.rb]: This one I actually would prefer to leave on. Basically because I don't want to be "that guy" writing spacing comments on people's PRs.

File: WcaOnRails/app/models/competition.rb [hunk 7]
```diff
@@ -513,7 +511,7 @@ def end_date
     if @start_date.blank?
       self.year = self.month = self.day = 0
     else
-      unless /\A\d{4}-\d{2}-\d{2}\z/.match(@start_date)
+      unless /\A\d{4}-\d{2}-\d{2}\z/ =~ @start_date
```
Comment (jfly) [inline WcaOnRails/app/models/competition.rb]: I'm okay turning this one off, it feels pretty silly to me. This is filed under [`Performance/RedundantMatch`](http://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Performance/RedundantMatch). I guess the idea is that `.match` does more work than `=~` building up some object that we don't really use. This feels like quite a microoptimization however.

I'll wait for you to respond before doing anything here, but I personally am inclined to turn off `Performance/RedundantMatch`.

File: WcaOnRails/app/models/competition.rb [hunk 8]
```diff
@@ -640,11 +637,11 @@ def in_progress?
     !results_posted? && (start_date..end_date).cover?(Date.today)
   end
 
-  # The name `is_probably_over` is meant to be surprising.
+  # The name `probably_over` is meant to be surprising.
   # We don't actually know when competitions are over, because we don't know their schedules, nor
   # do we know their timezones.
   # See discussion here: https://github.com/thewca/worldcubeassociation.org/pull/1206/files#r98485399.
-  def is_probably_over?
+  def probably_over?
```
Comment (jfly) [inline WcaOnRails/app/models/competition.rb]: Oh cool! I thought it was a ruby thing to never prefix with `is_*` because we have this suffix with `?` convention. I do think that the `is_` reads better most of the time, however, so I'm happy to revert this and turn off the cop.

File: WcaOnRails/app/controllers/competitions_controller.rb (hunk 3 above)
Comment (timhabermaas) [inline WcaOnRails/app/controllers/competitions_controller.rb]: I think there was a performance penalty for using double quoted strings compared to single quoted strings because they allow for interpolation, but this has been fixed in recent Ruby versions or at least the difference in performance isn't significant anymore: http://stackoverflow.com/questions/1836467/is-there-a-performance-gain-in-using-single-quotes-vs-double-quotes-in-ruby

👍 on ignoring this rule. Another disadvantage of using `''` everywhere is the added diff noise because you can neither use escape characters nor retroactively add interpolation without changing `'` to `"`.

File: WcaOnRails/app/models/post.rb [hunk 9]
```diff
@@ -29,25 +29,25 @@ def body_teaser
   CRASH_COURSE_POST_SLUG = "delegate-crash-course"
 
   def self.crash_course_post
-    post = ( Post.find_by_slug(CRASH_COURSE_POST_SLUG) ||
-             Post.create!(slug: CRASH_COURSE_POST_SLUG, title: "Delegate crash course", body: "Nothing here yet") )
+    post = (Post.find_by_slug(CRASH_COURSE_POST_SLUG) ||
+            Post.create!(slug: CRASH_COURSE_POST_SLUG, title: "Delegate crash course", body: "Nothing here yet"))
```
Comment (jfly) [inline WcaOnRails/app/models/post.rb]: Neat! I've created https://github.com/thewca/worldcubeassociation.org/issues/1299 and assigned it to myself. I'll take care of that in a seperate PR after this one.

File: WcaOnRails/app/models/user.rb [hunk 10]
```diff
@@ -568,7 +568,8 @@ def notify_of_id_claim_possibility(competition)
 
   def self.find_first_by_auth_conditions(warden_conditions)
     conditions = warden_conditions.dup
-    if login = conditions.delete(:login)
+    login = conditions.delete(:login)
```
Comment (jfly) [inline WcaOnRails/app/models/user.rb]: I'm also with Rubocop here, and do not respect the other camp =)

File: WcaOnRails/db/seeds/development/competitions.seeds.rb [hunk 11]
```diff
@@ -7,7 +7,7 @@ def random_events
     end
 
     def random_wca_value
-      r = rand(5000..100000)
+      r = rand(5000..100_000)
```
Comment (jfly) [inline WcaOnRails/db/seeds/development/competitions.seeds.rb]: Yeah, this is a pretty cool one!

File: WcaOnRails/db/migrate/20150504163657_create_doorkeeper_tables.rb [hunk 12]
```diff
@@ -2,38 +2,38 @@
...
-      t.string  :name,         null: false
-      t.string  :uid,          null: false
-      t.string  :secret,       null: false
-      t.text    :redirect_uri, null: false
-      t.string  :scopes,       null: false, default: ''
+      t.string :name, null: false
+      t.string :uid, null: false
+      t.string :secret, null: false
+      t.text :redirect_uri, null: false
+      t.string :scopes, null: false, default: ''
```
Comment (jfly) [inline WcaOnRails/db/migrate/20150504163657_create_doorkeeper_tables.rb]: This is kind of an interesting one. Rubocop actually *does* allow aligned code. However, this is in tension with their `Style/ExtraSpacing` rule, because without context, column aligned code looks like it has extra spaces. So I think what rubocop does is ignore the `Style/ExtraSpacing` rule if the next or previous lines have been spaced to align with the current line. This is problematic with the following code:

```ruby
 28     create_table :oauth_access_tokens do |t|
 29       t.integer  :resource_owner_id
 30       t.integer  :application_id
 31       t.string   :token,             null: false
 32       t.string   :refresh_token
 33       t.integer  :expires_in
 34       t.datetime :revoked_at
 35       t.datetime :created_at,        null: false
 36       t.string   :scopes
 37     end
```

```
~/gitting/worldcubeassociation.org/WcaOnRails @kaladin> bundle exec rubocop db/migrate/20150504163657_create_doorkeeper_tables.rb
Inspecting 1 file
C

Offenses:

db/migrate/20150504163657_create_doorkeeper_tables.rb:31:25: C: Unnecessary spacing detected.
      t.string   :token,             null: false
                        ^^^^^^^^^^^^
db/migrate/20150504163657_create_doorkeeper_tables.rb:35:30: C: Unnecessary spacing detected.
      t.datetime :created_at,        null: false
                             ^^^^^^^

1 file inspected, 2 offenses detected
~/gitting/worldcubeassociation.org/WcaOnRails @kaladin> 
```

However, if I reorder the lines with the `null: false` so they're touching, then the cop is happy:

```
 28     create_table :oauth_access_tokens do |t|
 29       t.integer  :resource_owner_id
 30       t.integer  :application_id
 31       t.string   :refresh_token      
 32       t.integer  :expires_in
 33       t.datetime :revoked_at
 34       t.datetime :created_at,        null: false
 35       t.string   :token,             null: false
 36       t.string   :scopes
 37     end
```

```
~/gitting/worldcubeassociation.org/WcaOnRails @kaladin> bundle exec rubocop db/migrate/20150504163657_create_doorkeeper_tables.rb
Inspecting 1 file
.

1 file inspected, no offenses detected
```

I've opted to reorder the lines here, rather than turning off the extra spacing cop.

File: WcaOnRails/db/seeds/development/competitions.seeds.rb [hunk 13]
```diff
@@ -70,8 +70,8 @@ def random_wca_value
             value3: random_wca_value,
             value4: random_wca_value,
             value5: random_wca_value,
-            regionalSingleRecord: i == 0 ? "WR" : "",
-            regionalAverageRecord: i == 0 ? "WR" : "",
+            regionalSingleRecord: j == 0 ? "WR" : "",
+            regionalAverageRecord: j == 0 ? "WR" : "",
```
Comment (jfly) [inline WcaOnRails/db/seeds/development/competitions.seeds.rb]: Why would you expect them to not be allowed here? There was something about using double quotes inside of interpolation inside of double quotes, but this is not related to that.

File: WcaOnRails/lib/solve_time.rb [hunk 14]
```diff
@@ -166,21 +166,21 @@ def clock_format
           result = "0:#{time_seconds}"
         else
           while time_seconds >= 60
-            result = ":%02d#{result}" % ( time_seconds % 60 )
-            time_seconds = time_seconds / 60
+            result = format(":%02d#{result}", time_seconds % 60)
```
Comment (jfly) [inline WcaOnRails/lib/solve_time.rb]: I'm also used to the `%` syntax, courtesy of python. However, I do think that it's a weird use of the character, so I like this rule.

**Comment by jfly** [comment]
I've updated this in light of the comments received so far. There are two open discussions:

- https://github.com/thewca/worldcubeassociation.org/pull/1286#discussion_r100163577
- https://github.com/thewca/worldcubeassociation.org/pull/1286#discussion_r100164631

File: WcaOnRails/app/controllers/polls_controller.rb (hunk 4 above)
Comment (larspetrus) [inline WcaOnRails/app/controllers/polls_controller.rb]: Yeah, "when in doubt, turn off the rule" is my approach.

There is a lot of air between "A is usually better than B" and "B should be forbidden". An informal team standard that you can break when it makes sense is my ideal.

File: WcaOnRails/app/models/user.rb (hunk 10 above)
Comment (larspetrus) [inline WcaOnRails/app/models/user.rb]: If this is how your team codes and you're used to seeing it, it streamlines some fairly common constructs.

But I'm not used to seeing it.

File: WcaOnRails/db/seeds/development/competitions.seeds.rb (hunk 13 above)
Comment (larspetrus) [inline WcaOnRails/db/seeds/development/competitions.seeds.rb]: Oh, _that's_ what was going on in `competitions_controller.rb`!

File: WcaOnRails/app/models/competition.rb (hunk 7 above)
Comment (larspetrus) [inline WcaOnRails/app/models/competition.rb]: Agreed.

**Comment by larspetrus** [comment]
LGTM, I guess.

File: WcaOnRails/app/models/competition.rb (hunk 7 above)
Comment (jfly) [inline WcaOnRails/app/models/competition.rb]: Done!

File: WcaOnRails/app/controllers/polls_controller.rb (hunk 4 above)
Comment (jfly) [inline WcaOnRails/app/controllers/polls_controller.rb]: I agree. I've disabled `Style/ConditionalAssignment`.

**Comment by jfly** [comment]
Updated in light of the latest reviews from @larspetrus. I'll leave this up for now, planning to merge it on Friday.
